Teach RemoteMirror how to project enum values

tbkka opened the pull request:
This adds two new functions to the SwiftRemoteMirror
facility that support inspecting enum values.

Currently, these support non-payload enums and
single-payload enums, including nested enums and
payloads with struct, tuple, and reference payloads.
In particular, it handles nested `Optional` types.

TODO: Multi-payload enums use different strategies for
encoding the cases that aren't yet supported by this
code.

Resolves rdar://59961527

```
/// Projects the value of an enum.
///
/// Takes the address and typeref for an enum and determines the
/// index of the currently-selected case within the enum.
///
/// Returns true iff the enum case could be successfully determined.
/// In particular, note that this code may fail for valid in-memory data
/// if the compiler is using a strategy we do not yet understand.
SWIFT_REMOTE_MIRROR_LINKAGE
int swift_reflection_projectEnumValue(SwiftReflectionContextRef ContextRef,
                                      swift_addr_t EnumAddress,
                                      swift_typeref_t EnumTypeRef,
                                      int *CaseIndex);

/// Finds information about a particular enum case.
///
/// Given an enum typeref and index of a case, returns:
/// * Typeref of the associated payload or zero if there is no payload
/// * Name of the case if known.
///
/// The Name points to a freshly-allocated C string on the heap.  You
/// are responsible for freeing the string (via `free()`) when you are finished.
SWIFT_REMOTE_MIRROR_LINKAGE
int swift_reflection_getEnumCaseTypeRef(SwiftReflectionContextRef ContextRef,
                                        swift_typeref_t EnumTypeRef,
                                        int CaseIndex,
                                        char **CaseName,
                                        swift_typeref_t *PayloadTypeRef);
```

mikeash committed:
[Reflection] Start roughing out a swift_reflection_projectEnum function.

rdar://problem/31666592

tbkka committed:
Merge remote-tracking branch 'github-mikeash/reflection-projectenum' into tbkka-remoteMirror-projectEnum

tbkka committed:
Teach RemoteMirror how to project enum values

This adds two new functions to the SwiftRemoteMirror
facility that support inspecting enum values.

Currently, these support non-payload enums and
single-payload enums, including nested enums and
payloads with struct, tuple, and reference payloads.
In particular, it handles nested `Optional` types.

TODO: Multi-payload enums use different strategies for
encoding the cases that aren't yet supported by this
code.

Resolves rdar://59961527

```
/// Projects the value of an enum.
///
/// Takes the address and typeref for an enum and determines the
/// index of the currently-selected case within the enum.
///
/// Returns true iff the enum case could be successfully determined.
/// In particular, note that this code may fail for valid in-memory data
/// if the compiler is using a strategy we do not yet understand.
SWIFT_REMOTE_MIRROR_LINKAGE
int swift_reflection_projectEnumValue(SwiftReflectionContextRef ContextRef,
                                      swift_addr_t EnumAddress,
                                      swift_typeref_t EnumTypeRef,
                                      uint64_t *CaseIndex);

/// Finds information about a particular enum case.
///
/// Given an enum typeref and index of a case, returns:
/// * Typeref of the associated payload or zero if there is no payload
/// * Name of the case if known.
///
/// The Name points to a freshly-allocated C string on the heap.  You
/// are responsible for freeing the string (via `free()`) when you are finished.
SWIFT_REMOTE_MIRROR_LINKAGE
int swift_reflection_getEnumCaseTypeRef(SwiftReflectionContextRef ContextRef,
                                        swift_typeref_t EnumTypeRef,
                                        unsigned CaseIndex,
                                        char **CaseName,
                                        swift_typeref_t *PayloadTypeRef);
```

tbkka commented:
This is not quite ready to merge:  It currently hard-codes some pointer properties that need to be queried from the target.

tbkka committed:
Move XI pointer configuration down to DLQ

Data Layout Queries (DLQ) are questions handled by queryDataLayout()
about the target environment.

This adds two new parameters:
  DLQ_GetLeastValidPointerValue
  DLQ_GetObjCReservedLowBits

These are the key parameters needed for the pointer XI
calculations.

TODO: The implementation of the new DLQ parameters is
still incomplete.

dcci commented on include/swift/Remote/InProcessMemoryReader.h:
```diff
@@ -30,18 +30,40 @@ class InProcessMemoryReader final : public MemoryReader {
+    case DLQ_GetObjCReservedLowBits: {
+      auto result = static_cast<uint8_t *>(outBuffer);
+#if __APPLE__ && __x86_64__
+      // ObjC on 64-bit macOS reserves the bottom bit of all pointers.
```
What about iOS?

tbkka replied:
As far as I understand, iOS Obj-C reserves high bits, not low bits.  So this literally only applies to 64-bit macOS, no other platforms.

dcci replied:
Thanks for clarifying.

tbkka committed:
Select pointer configuration based on compile-time settings.

This is insufficient for cases where you are inspecting an
image for another platform, but it's sufficient for the
test suites (which are always run on the same environment
as the target) and at least serves to document the necessary
logic for clients that need to support cross-platform inspection.

tbkka commented on include/swift/Reflection/TypeLowering.h:
```diff
@@ -155,6 +173,17 @@ class BuiltinTypeInfo : public TypeInfo {
+      return true;
+    }
+    // XXX Has extra inhabitants, so it must be a pointer?
+    return reader.readHeapObjectExtraInhabitantIndex(address, extraInhabitantIndex);
```
I'm not sure whether a `BuiltinTypeInfo` with extra inhabitants can ever be a function pointer or whether it's always just a regular pointer to heap data.  (There are different rules for managing XIs for function pointers.)

tbkka commented on include/swift/Reflection/TypeLowering.h:
```diff
@@ -206,6 +239,16 @@ class ReferenceTypeInfo : public TypeInfo {
+      *extraInhabitantIndex = -1;
+      return true;
+    }
+    return reader.readHeapObjectExtraInhabitantIndex(address, extraInhabitantIndex);
```
I suspect `ReferenceTypeInfo` is used for both function pointers and heap pointers, but I'm not quite sure how to distinguish those cases.

tbkka committed:
Rename file

tbkka committed:
Use Function ptr XI calculations for Thin Func ptrs

tbkka commented:
I think I've worked out the pointer properties as well as I can for now.  I'm still checking some details around XI handling for different kinds of pointers, but otherwise I think this is mergeable (pending CI, of course).

tbkka commented:
@swift-ci Please test macOS

tbkka commented:
@swift-ci Please test Linux

tbkka committed:
Fix finding the most-capacious field

The earlier version had a bad bug (a reused variable)
that ended up always using the last field instead of the
one with the most extra inhabitants.  Using STL algorithms
is safer.

tbkka committed:
Use 64-bit field for 8-byte pointers.

tbkka committed:
Fix spelling `wordSize` => `WordSize`

tbkka committed:
Move readExtraInhabitantIndex into the `.cpp` file

tbkka committed:
Fix use of Apple target macros

For some reason, `TARGET_OS_IOS` is defined to be 1 on iOS and defined to 0 on
all other Apple platforms.  So you can't just check whether the symbol is
defined or not.

It probably would suffice to just use the value, but I had headaches a while
back with some compiler that insisted on complaining about undefined
preprocessor macros, so I've used the horribly conservative `defined(FOO) &&
FOO` construction here.

tbkka committed:
Flesh out this test with 32-bit and 64-bit options, verify that multi-payload-enums are rejected (since they're not actually supported)

tbkka commented:
I think this is ready to merge.  The tests could be a little more comprehensive, but the API seems to be working correctly for the current tests on both macOS x86_64 and the 32-bit iOS simulator.

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please test macOS

tbkka committed:
Merge branch 'master' into tbkka-remoteMirror-projectEnum

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please test macOS

tbkka committed:
Merge branch 'master' into tbkka-remoteMirror-projectEnum

tbkka committed:
Fill in, simplify test cases

Add some missing 32-bit validation.

Also, when printing a payload-carrying enum value, just
include the type info ("struct foo") instead of the detailed
type ref.

tbkka commented:
@swift-ci Please test macOS

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please smoke test Linux

rintaro commented:
This caused test failure in "Swift in the OS" bot
https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RA-test-simulator-swift-in-the-os/1589/console
@tbkka Could you take a look? rdar://problem/60183719

tbkka commented:
PR #30281 should fix this.

slavapestov commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -710,6 +711,128 @@ class ReflectionContext
+    return false;
+  }
+
+  bool getEnumCaseTypeRef(const TypeRef *EnumTR,
```
I think this API is redundant. In the C API, we already have swift_reflection_childOfTypeRef()
